### PR TITLE
LUKS data moved to singleton

### PR DESCRIPTION
**Pull request by japokorn**
- added "static_data" directory
- the following was moved to luks_data singleton:
    - luks_devs
    - encryption_passphrase
    - luks_min_entropy
    - save_passphrase
    - min_luks_entropy attribute

### CODE REVIEW

File: blivet/autopart.py [hunk 1]
```diff
@@ -494,15 +486,17 @@ def do_autopart(storage, data, min_luks_entropy=0):
...
+    devs = _schedule_implicit_partitions(storage, disks)
     log.debug("candidate disks: %s", disks)
     log.debug("devs: %s", devs)
 
     if disks == []:
         raise NotEnoughFreeSpaceError(_("Not enough free space on disks for "
                                         "automatic partitioning"))
 
-    devs = _schedule_partitions(storage, disks, devs, min_luks_entropy=min_luks_entropy)
+    luks_data.min_entropy = min_luks_entropy
```
Comment (dwlehman) [inline blivet/autopart.py]: I don't think `min_luks_entropy` is defined here.

File: blivet/static_data/luks_data.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+
+class LUKS_Data(object):
+    """ Class to be used as a singleton.
+        Maintains the LUKS data.
+    """
+
+    def __init__(self):
+        """
+            TODO - fix doc text
```
Comment (vpodzime) [inline blivet/static_data/luks_data.py]: Would be nice to DO the TODO :)

File: blivet/static_data/luks_data.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    def save_passphrase(self, device):
+        """ Save a device's LUKS passphrase in case of reset. """
+        passphrase = device.format._LUKS__passphrase
+        if passphrase:
+            luks_data.devs[device.format.uuid] = passphrase
+            self.add_passphrase(passphrase)
+
+    @property
+    def passphrases(self):
+        return self. __passphrases
```
Comment (dwlehman) [inline blivet/static_data/luks_data.py]: Is there a space between `self.` and `__passphrases` here?

File: blivet/blivet.py [hunk 4]
```diff
@@ -145,8 +144,8 @@ def __init__(self, ksdata=None):
 
         # these will both be empty until our reset method gets called
         self.devicetree = DeviceTree(conf=self.config,
-                                     passphrase=self.encryption_passphrase,
-                                     luks_dict=self.__luks_devs)
+                                     passphrase=luks_data.encryption_passphrase,
+                                     luks_dict={})
```
Comment (vpodzime) [inline blivet/blivet.py]: I think the ``luks_dict`` parameter should default to ``None`` with ``DeviceTree`` creating the empty dict in such cases.

File: blivet/devicefactory.py [hunk 5]
```diff
@@ -336,7 +337,7 @@ def __init__(self, storage, size, disks, fstype=None, mountpoint=None,
 
         self.child_factory = None
         self.parent_factory = None
-        self.min_luks_entropy = min_luks_entropy
+        luks_data.min_entropy = min_luks_entropy
```
Comment (vpodzime) [inline blivet/devicefactory.py]: I'm not sure this should be set globally. It makes some sense to specify minimal entropy required per device. It makes sense in ``_schedule_partitions()`` and autopart in general, but for "explicitly created" devices it should be possible to specify minimal entropy separately.

Comment (vpodzime) [inline blivet/devicefactory.py]: And maybe store the value for future devices if it is not set explicitly again then.

File: blivet/devicefactory.py [hunk 6]
```diff
@@ -661,7 +662,7 @@ def _create_device(self):
 
             fmt = get_format(self.fstype,
                              mountpoint=self.mountpoint,
-                             min_luks_entropy=self.min_luks_entropy,
+                             min_luks_entropy=luks_data.min_entropy,
```
Comment (vpodzime) [inline blivet/devicefactory.py]: If it is, however, global, it should be passed as as an argument.

File: blivet/populator/populator.py [hunk 7]
```diff
@@ -73,6 +73,9 @@ def __init__(self, conf=None, passphrase=None, luks_dict=None):
             :type luks_dict: dict
 
         """
+
+        luks_data.luks_dict = luks_dict
```
Comment (vpodzime) [inline blivet/populator/populator.py]: This should be ``luks_data.devs`` I guess?

File: blivet/static_data/luks_data.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        self.__min_entropy = 0
+        self.__passphrases = []
+
+    @property
+    def devs(self):
+        return self.__devs
+
+    @devs.setter
+    def devs(self, value):
+        self.__devs = value
```
Comment (vpodzime) [inline blivet/static_data/luks_data.py]: I think ``devs`` is not a good name for this. Maybe it doesn't have to be publicly available like this at all? I believe all that's needed is a method that would return a passphrase for a given device.

**Comment by japokorn** [comment]
Uploaded new and generally more awesome version. As per comments:
@dwlehman:
 * `autopart.py`: `min_luks_entropy` is in that case defined as an optional method parameter
 * `luks_data.py`: Rogue space removed
@vpodzime:
 * `luks_data.py`: TODO removed, added comments
 * `blivet.py`, `devicefactory.py`: 
     * Kept to support backwards-compatibility
     * Will be changed/removed in blivet 3.0

**Comment by dwlehman** [comment]
I see now that you removed the docs for `min_luks_entropy` but not the actual argument handling. Why not use the global value like you're doing elsewhere, or at least use that as a default value for the kwarg?

File: blivet/autopart.py [hunk 9]
```diff
@@ -450,16 +445,13 @@ def do_reqpart(storage, requests):
...
+def do_autopart(storage, data, min_luks_entropy=None):
     """ Perform automatic partitioning.
 
         :param storage: a :class:`~.Blivet` instance
         :type storage: :class:`~.Blivet`
         :param data: kickstart data
         :type data: :class:`pykickstart.BaseHandler`
-        :param min_luks_entropy: minimum entropy in bits required for
-                                 luks format creation
-        :type min_luks_entropy: int
```
Comment (vpodzime) [inline blivet/autopart.py]: the docstring should stay just being updated with what ``None`` means

File: blivet/autopart.py [hunk 10]
```diff
@@ -494,15 +486,17 @@ def do_autopart(storage, data, min_luks_entropy=0):
...
-    devs = _schedule_implicit_partitions(storage, disks, min_luks_entropy)
+    devs = _schedule_implicit_partitions(storage, disks)
     log.debug("candidate disks: %s", disks)
     log.debug("devs: %s", devs)
 
     if disks == []:
         raise NotEnoughFreeSpaceError(_("Not enough free space on disks for "
                                         "automatic partitioning"))
+    if min_luks_entropy is not None:
+        luks_data.min_entropy = min_luks_entropy
```
Comment (vpodzime) [inline blivet/autopart.py]: I think this needs to happen before ``_schedule_implicit_partitions()`` is called.

File: blivet/deviceaction.py [hunk 11]
```diff
@@ -581,7 +582,7 @@ def execute(self, callbacks=None):
 
         if isinstance(self.device.format, luks.LUKS):
             # LUKS needs to wait for random data entropy if it is too low
-            min_required_entropy = self.device.format.min_luks_entropy
+            min_required_entropy = luks_data.min_entropy
```
Comment (vpodzime) [inline blivet/deviceaction.py]: I think this needs to check if ``format.min_luks_entropy`` is set or not and use the global value only if not.

File: blivet/formats/luks.py [hunk 12]
```diff
@@ -245,7 +238,7 @@ def _create(self, **kwargs):
                                     key_file=self._key_file,
                                     cipher=self.cipher,
                                     key_size=self.key_size,
-                                    min_entropy=self.min_luks_entropy)
+                                    min_entropy=luks_data.min_entropy)
```
Comment (vpodzime) [inline blivet/formats/luks.py]: This all needs to stay as is if we want to support per-device (more precisely per-format) entropy threshold.

File: blivet/autopart.py (hunk 9 above)
Comment (japokorn) [inline blivet/autopart.py]: Fixed.

File: blivet/autopart.py (hunk 10 above)
Comment (japokorn) [inline blivet/autopart.py]: Fixed.

File: blivet/deviceaction.py (hunk 11 above)
Comment (japokorn) [inline blivet/deviceaction.py]: Fixed.

File: blivet/formats/luks.py (hunk 12 above)
Comment (japokorn) [inline blivet/formats/luks.py]: Fixed.

File: blivet/autopart.py (hunk 9 above)
Comment (vpodzime) [inline blivet/autopart.py]: Thanks, but it should go next to the other parameters not to the block of text below. ;)

**Comment by vpodzime** [comment]
Looks good to me other than the nitpick above.

**Pull request merged**